### PR TITLE
fix(codepipeline): allow adding an S3 source action with the same bucket multiple times

### LINKED ISSUE #4237: Multiple S3SourceActions returns "already a Construct with name" error
When adding more than one S3SourceAction to a pipeline I get a duplicate Construct name error.

> jsii.errors.JSIIError: There is already a Construct with name 'joetestintegrationpipelinePipeline5832AB7ESourceEventRule' in Bucket [Bucket]

### Reproduction Steps

```python
from aws_cdk import (
    aws_codepipeline_actions as actions,
    aws_codepipeline as pipe,
    aws_s3 as s3,
    core
)

class MyStack(core.Stack):
    def __init__(self, scope: core.Construct, id: str, **kwargs) -> None:
        super().__init__(scope, id, **kwargs)

        pipeline = pipe.Pipeline(self, "Pipeline")
        pipe_bucket = s3.Bucket(self, "Bucket", versioned=True)
        action1 = actions.S3SourceAction(
            bucket=pipe_bucket,
            bucket_key='test1',
            output=pipe.Artifact(),
            action_name='test1',
            trigger=actions.S3Trigger.EVENTS
        )
        action2 = actions.S3SourceAction(
            bucket=pipe_bucket,
            bucket_key='test2',
            output=pipe.Artifact(),
            action_name='test2',
            trigger=actions.S3Trigger.EVENTS
        )

        source_stage = pipeline.add_stage(
            stage_name='Source',
            actions=[action1, action2])


```

### Error Log

```
jsii.errors.JavaScriptError: 
  Error: There is already a Construct with name 'joetestintegrationpipelinePipeline5832AB7ESourceEventRule' in Bucket [Bucket]
      at ConstructNode.addChild (/tmp/jsii-kernel-iAn6en/node_modules/@aws-cdk/core/lib/construct.js:420:19)
      at new ConstructNode (/tmp/jsii-kernel-iAn6en/node_modules/@aws-cdk/core/lib/construct.js:33:24)
      at new Construct (/tmp/jsii-kernel-iAn6en/node_modules/@aws-cdk/core/lib/construct.js:460:21)
      at new Resource (/tmp/jsii-kernel-iAn6en/node_modules/@aws-cdk/core/lib/resource.js:13:9)
      at new Rule (/tmp/jsii-kernel-iAn6en/node_modules/@aws-cdk/aws-events/lib/rule.js:14:9)
      at Bucket.onCloudTrailEvent (/tmp/jsii-kernel-iAn6en/node_modules/@aws-cdk/aws-s3/lib/bucket.js:49:22)
      at Bucket.onCloudTrailPutObject (/tmp/jsii-kernel-iAn6en/node_modules/@aws-cdk/aws-s3/lib/bucket.js:73:27)
      at S3SourceAction.bound (/tmp/jsii-kernel-iAn6en/node_modules/@aws-cdk/aws-codepipeline-actions/lib/s3/source-action.js:43:31)
      at S3SourceAction.bind (/tmp/jsii-kernel-iAn6en/node_modules/@aws-cdk/aws-codepipeline-actions/lib/action.js:23:21)
      at Kernel._wrapSandboxCode (/home/jlawson/src/codecommit/pipeline-cdk/.env/lib/python3.7/site-packages/jsii/_embedded/jsii/jsii-runtime.js:7253:19)
      at /home/jlawson/src/codecommit/pipeline-cdk/.env/lib/python3.7/site-packages/jsii/_embedded/jsii/jsii-runtime.js:6631:25
      at Kernel._ensureSync (/home/jlawson/src/codecommit/pipeline-cdk/.env/lib/python3.7/site-packages/jsii/_embedded/jsii/jsii-runtime.js:7226:20)
      at Kernel.invoke (/home/jlawson/src/codecommit/pipeline-cdk/.env/lib/python3.7/site-packages/jsii/_embedded/jsii/jsii-runtime.js:6630:26)
      at KernelHost.processRequest (/home/jlawson/src/codecommit/pipeline-cdk/.env/lib/python3.7/site-packages/jsii/_embedded/jsii/jsii-runtime.js:6327:28)
      at KernelHost.run (/home/jlawson/src/codecommit/pipeline-cdk/.env/lib/python3.7/site-packages/jsii/_embedded/jsii/jsii-runtime.js:6270:14)
      at Immediate._onImmediate (/home/jlawson/src/codecommit/pipeline-cdk/.env/lib/python3.7/site-packages/jsii/_embedded/jsii/jsii-runtime.js:6273:37)

```

### Environment

  - **CLI Version      :** 1.9.0 (build 30f158a)
  - **Framework Version:** 1.9.0
  - **OS               :** Arch Linux
  - **Language         :** Python 3.7

### Other

N/A

--- 

This is :bug: Bug Report

### PR DESCRIPTION
**Pull request by skinny85**
If you attempted to add two source S3 actions using the same bucket into the same pipeline, and both using CloudWatch events as triggers, you would get an error about 'duplicate construct with id PipelineSourceEventRule'. This is a valid use-case, though, as each action might point to a different path in the bucket.

Fixes #4237

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-codepipeline-actions/lib/s3/source-action.ts [hunk 1]
```diff
@@ -81,13 +81,34 @@ export class S3SourceAction extends Action {
...
+    }
+
     this.props = props;
   }
 
   protected bound(_scope: Construct, stage: codepipeline.IStage, options: codepipeline.ActionBindOptions):
       codepipeline.ActionConfig {
     if (this.props.trigger === S3Trigger.EVENTS) {
-      this.props.bucket.onCloudTrailPutObject(stage.pipeline.node.uniqueId + 'SourceEventRule', {
+      const baseId = stage.pipeline.node.uniqueId + 'SourceEventRule';
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-codepipeline-actions/lib/s3/source-action.ts]: I think the logic might be easier to follow if you just let `id` be mutable.

```ts
let id = ....;
if (props.bucket.node.tryFindChild(id) !== undefined) {
  id = ...;
}
```

Comment (skinny85) [reply packages/@aws-cdk/aws-codepipeline-actions/lib/s3/source-action.ts]: `id ` is already mutable: it's defined with `let id = ...;`.

**Comment by rix0rrr** [comment]
I think the logic of trying to keep the `Event::Rule` logical ID the same is not necessary; rules can be replaced without loss of service. Just switch over to the full new naming scheme.

We might trigger a pipeline twice during the cutover (which should be a window of seconds), but I think that's okay.

**Comment by skinny85** [comment]
> I think the logic of trying to keep the Event::Rule logical ID the same is not necessary; rules can be replaced without loss of service. Just switch over to the full new naming scheme.

Done in the newest revision.

**Pull request merged**